tools/amplify-preview/amplify.go

```diff
@@ -0,0 +1,338 @@
+/*
+Copyright 2024 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/aws/arn"
+	"github.com/aws/aws-sdk-go-v2/service/amplify"
+	"github.com/aws/aws-sdk-go-v2/service/amplify/types"
+)
+
+var (
+	errBranchNotFound           = errors.New("branch not found")
+	errNoJobForBranch           = errors.New("current branch has no jobs")
+	amplifyJobCompletedStatuses = map[types.JobStatus]struct{}{
+		types.JobStatusFailed:    {},
+		types.JobStatusCancelled: {},
+		types.JobStatusSucceed:   {},
+	}
+)
+
+const (
+	logKeyAppID      = "appID"
+	logKeyBranchName = "branchName"
+	logKeyJobID      = "jobID"
+
+	amplifyMarkdownHeader = "Amplify deployment status"
+	amplifyDefaultDomain  = "amplifyapp.com"
+)
+
+type AmplifyPreview struct {
+	appIDs     []string
+	branchName string
+	client     *amplify.Client
+}
+
+type aggregatedError struct {
+	perAppErr map[string]error
+	message   string
+}
+
+func (amp *AmplifyPreview) FindExistingBranch(ctx context.Context) (*types.Branch, error) {
+	type resp struct {
+		appID string
+		data  *amplify.GetBranchOutput
+		err   error
+	}
+	var wg sync.WaitGroup
+	wg.Add(len(amp.appIDs))
+	resultCh := make(chan resp, len(amp.appIDs))
+
+	for _, appID := range amp.appIDs {
+		go func() {
+			defer wg.Done()
+			branch, err := amp.client.GetBranch(ctx, &amplify.GetBranchInput{
+				AppId:      aws.String(appID),
+				BranchName: aws.String(amp.branchName),
+			})
+			resultCh <- resp{
+				appID: appID,
+				data:  branch,
+				err:   err,
+			}
+		}()
+	}
+
+	wg.Wait()
+	close(resultCh)
+
+	failedResp := aggregatedError{
+		perAppErr: map[string]error{},
+		message:   "failed to fetch branch",
+	}
+
+	for resp := range resultCh {
+		var errNotFound *types.NotFoundException
+		if errors.As(resp.err, &errNotFound) {
+			logger.Debug("Branch not found", logKeyAppID, resp.appID, logKeyBranchName, amp.branchName)
+			continue
+		} else if resp.err != nil {
+			failedResp.perAppErr[resp.appID] = resp.err
+			continue
+		}
+
+		if resp.data != nil {
+			return resp.data.Branch, nil
+		}
```

Comment:
It should be always just one branch, but we don't know which app it's connected to.
We totally can iterate without goroutines, just thought this will save save second or two, so we don't have to wait on sequential responses.